File: ios/BlueViewWeather/Update/UpdateChecker.swift
import Foundation

enum UpdateState: Equatable {
    case idle
    case checking
    case upToDate
    case available(version: String)
    case failed(String)
}

private struct GitHubRelease: Decodable {
    let tagName: String
    let draft: Bool?
    let prerelease: Bool?

    enum CodingKeys: String, CodingKey {
        case tagName = "tag_name"
        case draft
        case prerelease
    }
}

/// Mirrors the Android UpdateChecker, minus the download/install half: iOS cannot
/// self-install, so the newest ios-v tag only routes the user to TestFlight.
@MainActor
final class UpdateChecker: ObservableObject {
    @Published private(set) var state: UpdateState = .idle

    // The monorepo releases all three platforms on their own tag prefixes, so
    // /releases/latest is unusable here — it could return an android- or desktop- tag.
    private static let releasesURL =
        "https://api.github.com/repos/frank-blueview-ai/Blue-View-Ai-Weather/releases?per_page=30"
    private static let tagPrefix = "ios-v"

    static let testFlightURL = URL(string: "itms-beta://testflight.apple.com/")!
    static let testFlightFallbackURL = URL(string: "https://apps.apple.com/app/testflight/id899247664")!

    static var currentVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
    }

    private let session: URLSession
    private let decoder = JSONDecoder()

    init(session: URLSession = .shared) {
        self.session = session
    }

    func check() async {
        state = .checking

        var request = URLRequest(url: URL(string: Self.releasesURL)!)
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            state = .failed("Network error: \(error.localizedDescription)")
            return
        }

        guard let http = response as? HTTPURLResponse else {
            state = .failed("Unexpected server response")
            return
        }
        guard (200..<300).contains(http.statusCode) else {
            // 403 here is almost always the unauthenticated hourly rate limit.
            state = .failed(http.statusCode == 403
                ? "GitHub rate limit reached — try again later"
                : "Update check failed (HTTP \(http.statusCode))")
            return
        }

        let releases: [GitHubRelease]
        do {
            releases = try decoder.decode([GitHubRelease].self, from: data)
        } catch {
            state = .failed("Failed to parse release list")
            return
        }

        // The API's ordering is not guaranteed to be semver order, so pick the max.
        let versions = releases
            .filter { $0.tagName.hasPrefix(Self.tagPrefix) && $0.draft != true && $0.prerelease != true }
            .map { String($0.tagName.dropFirst(Self.tagPrefix.count)) }

        guard let latest = versions.max(by: { Self.isNewer($1, than: $0) }) else {
            state = .failed("No iOS release found")
            return
        }

        state = Self.isNewer(latest, than: Self.currentVersion) ? .available(version: latest) : .upToDate
    }

    /// Numeric per-component compare — "1.0.10" is newer than "1.0.9". Non-numeric
    /// components count as 0 so malformed tags degrade instead of throwing.
    static func isNewer(_ candidate: String, than current: String) -> Bool {
        let lhs = components(of: candidate)
        let rhs = components(of: current)
        for i in 0..<max(lhs.count, rhs.count) {
            let l = i < lhs.count ? lhs[i] : 0
            let r = i < rhs.count ? rhs[i] : 0
            if l != r { return l > r }
        }
        return false
    }

    private static func components(of version: String) -> [Int] {
        version
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .drop(while: { $0 == "v" || $0 == "V" })
            .split(separator: ".")
            .map { Int($0.prefix(while: { $0.isNumber })) ?? 0 }
    }
}

File: android/app/src/github/java/ai/blueview/weather/data/update/UpdateChecker.kt
package ai.blueview.weather.data.update

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import javax.inject.Inject
import javax.inject.Singleton

// github flavour only. The Play artifact must not contain any of this: Play's
// Device and Network Abuse policy forbids an app it distributes from updating
// itself outside Play or pointing users at another download of the same app.
// See src/play/.../UpdateChecker.kt for the stub that replaces it.

// The monorepo releases all three platforms, each on its own tag prefix, so
// /releases/latest is not usable here — it could return an ios- or desktop- tag.
// Fetch the list (newest first) and pick the newest android- release.
private const val RELEASES_API =
    "https://api.github.com/repos/frank-blueview-ai/Blue-View-Ai-Weather/releases?per_page=30"

private const val TAG_PREFIX = "android-v"

@Serializable
data class GithubRelease(
    @SerialName("tag_name") val tagName: String,
    @SerialName("assets")   val assets: List<GithubAsset>
)

@Serializable
data class GithubAsset(
    @SerialName("name")                 val name: String,
    @SerialName("browser_download_url") val downloadUrl: String
)

@Singleton
class UpdateChecker @Inject constructor(
    private val okHttpClient: OkHttpClient
) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun checkLatest(currentVersion: String): UpdateState = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url(RELEASES_API)
                .header("Accept", "application/vnd.github+json")
                .build()
            val body = okHttpClient.newCall(request).execute().use { it.body?.string() }
                ?: return@withContext UpdateState.Error("Empty response from server")
            val release = json.decodeFromString<List<GithubRelease>>(body)
                .firstOrNull { it.tagName.startsWith(TAG_PREFIX) }
                ?: return@withContext UpdateState.Error("No Android release found")
            val latest  = release.tagName.removePrefix(TAG_PREFIX)
            val current = currentVersion.trimStart('v')
            if (isNewer(latest, current)) {
                val apk = release.assets.firstOrNull { it.name.endsWith(".apk") }
                    ?: return@withContext UpdateState.Error("No APK found in release")
                UpdateState.Available(latest, apk.downloadUrl)
            } else {
                UpdateState.UpToDate
            }
        } catch (e: Exception) {
            UpdateState.Error(e.message ?: "Check failed")
        }
    }

    private fun isNewer(latest: String, current: String): Boolean {
        val l = latest.split(".").mapNotNull { it.toIntOrNull() }
        val c = current.split(".").mapNotNull { it.toIntOrNull() }
        for (i in 0 until maxOf(l.size, c.size)) {
            val lv = l.getOrElse(i) { 0 }
            val cv = c.getOrElse(i) { 0 }
            if (lv != cv) return lv > cv
        }
        return false
    }
}
